claragenomics/dl4atac/dataset.py

```diff
@@ -170,24 +170,31 @@ def _get_generator(self):
         for i, filename in enumerate(self.files):
             hf = h5py.File(filename, 'r')
             hd = hf["input"]
-            # Read additional layers
-            if self.layers is not None:
-                for layer_key in self.layers:
-                    hd = np.dstack((hd, hf[layer_key]))
             hdrecs.append(hd)
             sys.stdout.flush()
         idx = yield
         while True:
             # Find correct dataset, given idx
             file_id, local_idx = self._get_file_id(idx)
+            filename = self.files[file_id]
             assert file_id < len(self.files)
             if (local_idx < self.fh_indices[file_id][0]) or (
                     local_idx >= self.fh_indices[file_id][1]):
                 # Data is not pre loaded, so load new data
+                # Read additional layers
+                hf = h5py.File(filename, 'r')
+                prefetch_hdrecs = hdrecs[file_id][
```

Comment:
since the other layers are still being loaded from a file here, I don't see much of a point in loading the `input` layer in line 172 anymore. It's probably best to refactor it as the following - 
1. near line 170, just open the h5py file and store the file pointer (so we don't have to keep closing and opening it), and also update the "layers" list to have input + custom layer
2. near line 184, porefetch all layers and stack them up

Reply:
In line 170, if I do 
```
hdrecs.append(hf)```
instead of 
```
 hdrecs.append(hf["input"])```

program still crashes with OOM. I'm not sure how to save file pointer, as reading the h5 file stores the entire file object. Am I mis-understanding something ?

Reply:
Strange... In that case I'd suggest we just move the whole File reading down into line 184 as well, and take out the loop that's there near line 170. So 184 would look like this
```
hf = h5py.File(self.files[file_id], 'r')
prefetch_hdrec = hf[input]local_index + prefetch_size]
for layer in layers:
    prefetch_hdrec[layers] = np.stackprefetch_hdrec , (hf[layer]local_index + prefetch_size])
```